File: bench/tcp_server.go
// BOT-GENERATED

package main

import (
	"io"
	"net"
)

const ready = byte('R')
const data = byte('D')
const closeConnection = byte('Q')

func handle(conn net.Conn) {
	defer conn.Close()
	if _, err := conn.Write([]byte{ready}); err != nil {
		panic(err)
	}
	var buf [1]byte
	for {
		if _, err := io.ReadFull(conn, buf[:]); err != nil {
			panic(err)
		}
		if buf[0] != data && buf[0] != closeConnection {
			panic("bad data byte")
		}
		if _, err := conn.Write(buf[:]); err != nil {
			panic(err)
		}
		if buf[0] == closeConnection {
			return
		}
	}
}

func main() {
	listener, err := net.Listen("tcp4", "127.0.0.1:19777")
	if err != nil {
		panic(err)
	}

	for {
		conn, err := listener.Accept()
		if err != nil {
			panic(err)
		}
		go handle(conn)
	}
}

File: bench/tcp_server.java
// BOT-GENERATED

import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

final class tcp_server {
  private static final int READY = 'R';
  private static final int DATA = 'D';
  private static final int PORT = 19777;

  private static void handle(Socket socket) {
    try (socket) {
      final var out = socket.getOutputStream();
      final var in = socket.getInputStream();
      out.write(READY);
      if (in.read() != DATA) throw new AssertionError("bad DATA byte");
      out.write(DATA);
    } catch (Exception error) {
      throw new RuntimeException(error);
    }
  }

  static void run(ExecutorService executor) throws Exception {
    try (var listener = new ServerSocket(); executor) {
      listener.setReuseAddress(true);
      listener.bind(
        new InetSocketAddress(InetAddress.getLoopbackAddress(), PORT)
      );
      while (true) {
        final Socket socket = listener.accept();
        executor.execute(() -> handle(socket));
      }
    }
  }

  public static void main(String[] args) throws Exception {
    run(Executors.newVirtualThreadPerTaskExecutor());
  }
}

final class tcp_server_thread {
  public static void main(String[] args) throws Exception {
    tcp_server.run(
      Executors.newThreadPerTaskExecutor(Thread.ofPlatform().factory())
    );
  }
}
